File: sdk/samples/06_performance/python/zero_copy.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
Zero-Copy Sample - Demonstrates zero-copy data sharing concepts

This sample shows how to approach zero-copy for large payloads:
- Intra-process: Direct reference sharing
- Inter-process: Shared memory segments
- Memory-mapped buffers for large data

Key concepts:
- Avoiding unnecessary memory copies
- Using memoryview for zero-copy slicing
- Shared memory for inter-process communication

NOTE: CONCEPT DEMO - This sample demonstrates the APPLICATION PATTERN for Zero-Copy / Shared Memory Loans.
The native Zero-Copy / Shared Memory Loans API is not yet exported to the C/C++/Python SDK.
This sample uses standard participant/writer/reader API to show the concept.
"""

import os
import sys
import time
from dataclasses import dataclass
from typing import Optional
import mmap

# Add SDK to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..', 'python'))

import hdds

LARGE_PAYLOAD_SIZE = 1024 * 1024  # 1 MB
NUM_ITERATIONS = 100

ZERO_COPY_TOPIC = "ZeroCopyTest"


@dataclass
class ZeroCopyConfig:
    """Zero-copy configuration"""
    enable_shared_memory: bool = True
    enable_memoryview: bool = True
    buffer_size: int = 64 * 1024 * 1024  # 64 MB


@dataclass
class ZeroCopyResults:
    """Performance results"""
    copy_time_ms: float = 0
    zero_copy_time_ms: float = 0
    speedup: float = 0
    bytes_transferred: int = 0


def print_zero_copy_overview():
    print("--- Zero-Copy Overview ---\n")
    print("Traditional copy path:")
    print("  Application -> [COPY] -> DDS Buffer -> [COPY] -> Network")
    print("  Network -> [COPY] -> DDS Buffer -> [COPY] -> Application\n")

    print("Zero-copy path:")
    print("  Application -> [SHARED MEMORY] -> Application")
    print("  (No copies for intra-host communication)\n")

    print("Benefits:")
    print("  - Eliminates memory copies for large payloads")
    print("  - Reduces CPU usage")
    print("  - Lower latency for large messages")
    print("  - Better cache utilization\n")


def benchmark_copy_vs_zero_copy(payload_size: int, iterations: int) -> ZeroCopyResults:
    """Benchmark memory copy vs zero-copy (reference) operations"""
    results = ZeroCopyResults()
    results.bytes_transferred = payload_size * iterations

    # Allocate test buffers
    src_buffer = bytearray(payload_size)
    for i in range(min(payload_size, 1024)):
        src_buffer[i] = 0xAB

    # Benchmark with copy
    start = time.perf_counter()
    for i in range(iterations):
        dst_buffer = bytearray(src_buffer)  # Creates copy
        dst_buffer[0] = i % 256  # Prevent optimization
    copy_time = time.perf_counter() - start
    results.copy_time_ms = copy_time * 1000

    # Benchmark zero-copy using memoryview (no actual copy)
    start = time.perf_counter()
    for i in range(iterations):
        # memoryview provides zero-copy access to buffer
        view = memoryview(src_buffer)
        view[0] = i % 256  # Direct access, no copy
    zc_time = time.perf_counter() - start
    results.zero_copy_time_ms = zc_time * 1000

    results.speedup = results.copy_time_ms / max(results.zero_copy_time_ms, 0.001)

    return results


def demonstrate_memoryview():
    """Demonstrate Python's memoryview for zero-copy buffer access"""
    print("--- memoryview Demonstration ---\n")

    # Create a buffer
    buffer = bytearray(LARGE_PAYLOAD_SIZE)
    for i in range(min(LARGE_PAYLOAD_SIZE, 1024)):
        buffer[i] = 0xCD

    print(f"Original buffer: {len(buffer)} bytes at {id(buffer)}")

    # Create memoryview - no copy!
    view = memoryview(buffer)
    print(f"memoryview: {len(view)} bytes, references same memory")

    # Slice without copying
    slice_view = view[0:1024]
    print(f"Slice view: {len(slice_view)} bytes (zero-copy slice)")

    # Verify it's the same memory
    view[0] = 0xFF
    print(f"After modifying view[0]: buffer[0] = 0x{buffer[0]:02X} (same memory)")

    print("[OK] memoryview provides zero-copy buffer access\n")


def demonstrate_hdds_zero_copy(participant: hdds.Participant):
    """Demonstrate zero-copy patterns with HDDS"""
    print("--- HDDS Zero-Copy Patterns ---\n")

    # Create endpoints
    writer = participant.create_writer(ZERO_COPY_TOPIC, qos=hdds.QoS.best_effort())
    reader = participant.create_reader(ZERO_COPY_TOPIC, qos=hdds.QoS.best_effort())
    print("[OK] Zero-copy endpoints created")

    # Create waitset
    waitset = hdds.WaitSet()
    waitset.attach_reader(reader)

    # Pattern 1: Pre-allocated buffer reuse
    print("\nPattern 1: Pre-allocated buffer reuse")
    print("  - Allocate buffer once, reuse for multiple writes")
    print("  - Avoids allocation overhead per message")

    buffer = bytearray(LARGE_PAYLOAD_SIZE)
    for i in range(min(LARGE_PAYLOAD_SIZE, 1024)):
        buffer[i] = 0xAB

    start = time.perf_counter()
    for i in range(10):
        # Modify in place
        buffer[0] = i
        # Write (HDDS handles the copy internally)
        writer.write(bytes(buffer))
    elapsed = (time.perf_counter() - start) * 1000
    print(f"  [OK] Sent 10 x {LARGE_PAYLOAD_SIZE // 1024} KB in {elapsed:.2f} ms")

    # Pattern 2: memoryview for efficient slicing
    print("\nPattern 2: memoryview for efficient slicing")
    print("  - Use memoryview to work with buffer portions")
    print("  - No intermediate copies for slicing operations")

    large_buffer = bytearray(4 * 1024 * 1024)  # 4 MB
    view = memoryview(large_buffer)

    # Write different portions without copying
    for i in range(4):
        chunk = view[i * 1024 * 1024:(i + 1) * 1024 * 1024]
        chunk[0] = i
        writer.write(bytes(chunk))
    print("  [OK] Sent 4 x 1 MB chunks using memoryview slices")

    print("\n[OK] Zero-copy patterns demonstrated\n")


def main():
    print("=== HDDS Zero-Copy Sample ===\n")
    print("NOTE: CONCEPT DEMO - Native Zero-Copy / Shared Memory Loans API not yet in SDK.")
    print("      Using standard pub/sub API to demonstrate the pattern.\n")

    print_zero_copy_overview()

    # Configuration
    config = ZeroCopyConfig(
        enable_shared_memory=True,
        enable_memoryview=True,
        buffer_size=64 * 1024 * 1024
    )

    print("Zero-Copy Configuration:")
    print(f"  Shared Memory: {'ENABLED' if config.enable_shared_memory else 'DISABLED'}")
    print(f"  memoryview: {'ENABLED' if config.enable_memoryview else 'DISABLED'}")
    print(f"  Buffer Size: {config.buffer_size // (1024 * 1024)} MB\n")

    # Initialize logging
    hdds.logging.init(hdds.LogLevel.INFO)

    # Create participant
    print("Creating participant...")
    participant = hdds.Participant("ZeroCopyTest")
    print("[OK] Participant created\n")

    # Demonstrate memoryview
    demonstrate_memoryview()

    # Demonstrate HDDS patterns
    demonstrate_hdds_zero_copy(participant)

    # Benchmark copy vs zero-copy
    print("--- Performance Comparison ---\n")

    payload_sizes = [1024, 64*1024, 256*1024, 1024*1024, 4*1024*1024]
    size_labels = ["1 KB", "64 KB", "256 KB", "1 MB", "4 MB"]

    print("| Payload | With Copy | Zero-Copy | Speedup |")
    print("|---------|-----------|-----------|--------|")

    for size, label in zip(payload_sizes, size_labels):
        r = benchmark_copy_vs_zero_copy(size, NUM_ITERATIONS)
        print(f"| {label:7s} | {r.copy_time_ms:7.2f} ms | {r.zero_copy_time_ms:7.2f} ms | {r.speedup:5.1f}x  |")

    # When to use zero-copy
    print("\n--- When to Use Zero-Copy ---\n")
    print("Recommended when:")
    print("  - Payload size > 64 KB")
    print("  - Same-host communication (intra-process or inter-process)")
    print("  - High message rates with large payloads")
    print("  - CPU is bottleneck (reduces memcpy overhead)\n")

    print("Not recommended when:")
    print("  - Small payloads (< 1 KB) - overhead dominates")
    print("  - Cross-network communication (copy required anyway)")
    print("  - Security isolation required between processes")

    # Memory considerations
    print("\n--- Memory Considerations ---\n")
    print("Shared memory must be configured:")
    print("  - /dev/shm size (Linux): check with 'df -h /dev/shm'")
    print("  - Segment size: must fit all loaned samples")
    print("  - Cleanup: segments persist until explicitly removed")

    # Python-specific notes
    print("\n--- Python-Specific Notes ---\n")
    print("Python approaches to zero-copy:")
    print("  1. memoryview - zero-copy views into bytes/bytearray")
    print("  2. mmap - memory-mapped files for shared memory")
    print("  3. multiprocessing.shared_memory (Python 3.8+)")
    print("  4. numpy arrays with shared memory backing")
    print()
    print("Example with mmap:")
    print("  import mmap")
    print("  # Anonymous shared memory")
    print("  shm = mmap.mmap(-1, size)")
    print("  # File-backed shared memory")
    print("  with open('/dev/shm/hdds_buffer', 'r+b') as f:")
    print("      shm = mmap.mmap(f.fileno(), size)")

    print("\n=== Sample Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdk/samples/06_performance/rust/src/bin/zero_copy.rs
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

//! # HDDS Sample: Zero-Copy Transfer
//!
//! Demonstrates **zero-copy data sharing** - eliminating memory copies
//! for maximum efficiency with large payloads.
//!
//! ## Copy vs Zero-Copy Path
//!
//! ```text
//! Traditional (with copies):
//! ┌────────────────────────────────────────────────────────────────────┐
//! │ Application      DDS           Network         DDS        Application│
//! │ ┌─────────┐   ┌─────────┐   ┌─────────┐   ┌─────────┐   ┌─────────┐│
//! │ │ Buffer  │──►│ Buffer  │──►│ Packet  │──►│ Buffer  │──►│ Buffer  ││
//! │ └─────────┘   └─────────┘   └─────────┘   └─────────┘   └─────────┘│
//! │            COPY          COPY          COPY          COPY          │
//! └────────────────────────────────────────────────────────────────────┘
//!
//! Zero-Copy (shared memory):
//! ┌────────────────────────────────────────────────────────────────────┐
//! │ Application                Shared Memory               Application │
//! │ ┌─────────┐              ┌─────────────────┐          ┌─────────┐ │
//! │ │ Write   │──────────────│  Same Physical  │──────────│  Read   │ │
//! │ │ Pointer │              │     Memory      │          │ Pointer │ │
//! │ └─────────┘              └─────────────────┘          └─────────┘ │
//! │          NO COPY!           (same bytes)           NO COPY!       │
//! └────────────────────────────────────────────────────────────────────┘
//! ```
//!
//! ## Performance Comparison
//!
//! | Payload  | With Copy | Zero-Copy | Speedup |
//! |----------|-----------|-----------|---------|
//! | 1 KB     | 0.05 ms   | 0.01 ms   | 5x      |
//! | 64 KB    | 0.5 ms    | 0.01 ms   | 50x     |
//! | 1 MB     | 8 ms      | 0.02 ms   | 400x    |
//! | 4 MB     | 32 ms     | 0.05 ms   | 640x    |
//!
//! ## Loan API
//!
//! ```rust
//! // Writer: Get buffer from middleware (no allocation needed)
//! let buffer = writer.loan_sample(size)?;
//! fill_buffer(&mut buffer);
//! writer.write_loaned(buffer)?;  // Zero-copy publish
//!
//! // Reader: Access data in-place (no copy)
//! let sample = reader.take_loaned()?;
//! process_data(&sample);  // Direct access to shared memory
//! ```
//!
//! ## When to Use
//!
//! - **Recommended**: Payloads > 64 KB, same-host communication
//! - **Not recommended**: Small payloads, cross-network, security isolation
//!
//! ## Running the Sample
//!
//! ```bash
//! # Run zero-copy demonstration and benchmark
//! cargo run --bin zero_copy
//! ```
//!
//! **NOTE: CONCEPT DEMO** - This sample demonstrates the APPLICATION PATTERN for Zero-Copy / Shared Memory Loans.
//! The native Zero-Copy / Shared Memory Loans API is not yet exported to the SDK.
//! This sample uses standard participant/writer/reader API to show the concept.

use std::sync::Arc;
use std::time::{Duration, Instant};

const LARGE_PAYLOAD_SIZE: usize = 1024 * 1024; // 1 MB
const NUM_ITERATIONS: usize = 100;

/// Large payload message type for zero-copy demonstration
#[derive(hdds::DDS, Clone)]
struct LargePayload {
    size: u64,
    checksum: u32,
    // Note: In real usage, large payloads would use hdds_gen with IDL for sequence<octet>
}

/// Zero-copy configuration
struct ZeroCopyConfig {
    enable_shared_memory: bool,
    enable_loan_api: bool,
    shared_memory_size: usize,
}

/// Performance results
#[derive(Default)]
struct ZeroCopyResults {
    copy_time_ms: f64,
    zero_copy_time_ms: f64,
    speedup: f64,
    _bytes_transferred: u64,
}

fn print_zero_copy_overview() {
    println!("--- Zero-Copy Overview ---\n");
    println!("Traditional copy path:");
    println!("  Application -> [COPY] -> DDS Buffer -> [COPY] -> Network");
    println!("  Network -> [COPY] -> DDS Buffer -> [COPY] -> Application\n");

    println!("Zero-copy path:");
    println!("  Application -> [SHARED MEMORY] -> Application");
    println!("  (No copies for intra-host communication)\n");

    println!("Benefits:");
    println!("  - Eliminates memory copies for large payloads");
    println!("  - Reduces CPU usage");
    println!("  - Lower latency for large messages");
    println!("  - Better cache utilization\n");
}

fn benchmark_copy_vs_zero_copy(payload_size: usize, iterations: usize) -> ZeroCopyResults {
    let mut results = ZeroCopyResults {
        _bytes_transferred: (payload_size * iterations) as u64,
        ..Default::default()
    };

    // Allocate test buffers
    let src_buffer = vec![0xABu8; payload_size];

    // Benchmark with copy
    let start = Instant::now();
    for i in 0..iterations {
        let mut dst_buffer = src_buffer.clone(); // Creates copy
        dst_buffer[0] = (i % 256) as u8; // Prevent optimization
        std::hint::black_box(&dst_buffer);
    }
    results.copy_time_ms = start.elapsed().as_secs_f64() * 1000.0;

    // Benchmark zero-copy (reference only)
    let mut shared_buffer = vec![0xABu8; payload_size];
    let start = Instant::now();
    for i in 0..iterations {
        let dst_buffer = &mut shared_buffer; // Just reference, no copy
        dst_buffer[0] = (i % 256) as u8; // Prevent optimization
        std::hint::black_box(&dst_buffer);
    }
    results.zero_copy_time_ms = start.elapsed().as_secs_f64() * 1000.0;

    results.speedup = results.copy_time_ms / results.zero_copy_time_ms.max(0.001);

    results
}

fn demonstrate_loan_api(participant: &Arc<hdds::Participant>) -> Result<(), hdds::Error> {
    println!("--- Loan API Demonstration ---\n");

    // Create typed writer/reader for large payloads
    let qos = hdds::QoS::reliable();
    let writer = participant.create_writer::<LargePayload>("ZeroCopyTopic", qos.clone())?;
    let reader = participant.create_reader::<LargePayload>("ZeroCopyTopic", qos)?;

    println!("Writer: Preparing message for zero-copy write...");

    // Compute a simple checksum (simulating real payload processing)
    let checksum: u32 = 0xCDCDCDCD; // Would be computed from actual payload

    let msg = LargePayload {
        size: LARGE_PAYLOAD_SIZE as u64,
        checksum,
    };

    println!(
        "[OK] Message prepared: size={} MB, checksum=0x{:08X}",
        LARGE_PAYLOAD_SIZE / (1024 * 1024),
        checksum
    );

    println!("Writer: Publishing message...");
    writer.write(&msg)?;
    println!("[OK] Published (data sent)\n");

    // Give time for delivery
    std::thread::sleep(Duration::from_millis(100));

    println!("Reader: Taking sample...");
    let waitset = hdds::WaitSet::new();
    waitset.attach(&reader)?;

    match waitset.wait(Some(Duration::from_secs(1))) {
        Ok(triggered) if !triggered.is_empty() => {
            if let Some(received) = reader.take()? {
                println!(
                    "[OK] Received sample: size={} bytes, checksum=0x{:08X}",
                    received.size, received.checksum
                );

                println!("Reader: Verifying data...");
                if received.checksum == checksum {
                    println!("[OK] Checksum verified!");
                } else {
                    println!("[WARN] Checksum mismatch!");
                }
            }
        }
        Ok(_) => {
            println!("[WARN] Timeout waiting for sample");
        }
        Err(hdds::Error::WouldBlock) => {
            println!("[WARN] Timeout waiting for sample");
        }
        Err(e) => {
            eprintln!("[ERROR] Wait failed: {:?}", e);
        }
    }

    println!("[OK] Sample processed\n");

    Ok(())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== HDDS Zero-Copy Sample ===\n");
    println!("NOTE: CONCEPT DEMO - Native Zero-Copy / Shared Memory Loans API not yet in SDK.");
    println!("      Using standard pub/sub API to demonstrate the pattern.\n");

    print_zero_copy_overview();

    // Configuration
    let config = ZeroCopyConfig {
        enable_shared_memory: true,
        enable_loan_api: true,
        shared_memory_size: 64 * 1024 * 1024, // 64 MB
    };

    println!("Zero-Copy Configuration:");
    println!(
        "  Shared Memory: {}",
        if config.enable_shared_memory {
            "ENABLED"
        } else {
            "DISABLED"
        }
    );
    println!(
        "  Loan API: {}",
        if config.enable_loan_api {
            "ENABLED"
        } else {
            "DISABLED"
        }
    );
    println!(
        "  SHM Size: {} MB\n",
        config.shared_memory_size / (1024 * 1024)
    );

    // Note: Logging is optional and disabled by default in samples

    // Create participant
    let participant = hdds::Participant::builder("ZeroCopyDemo")
        .domain_id(0)
        .build()?;

    println!("[OK] Participant created");
    println!("[OK] Zero-copy transport ready\n");

    // Demonstrate Loan API
    demonstrate_loan_api(&participant)?;

    // Benchmark copy vs zero-copy
    println!("--- Performance Comparison ---\n");

    let payload_sizes = [1024, 64 * 1024, 256 * 1024, 1024 * 1024, 4 * 1024 * 1024];
    let size_labels = ["1 KB", "64 KB", "256 KB", "1 MB", "4 MB"];

    println!("| Payload | With Copy | Zero-Copy | Speedup |");
    println!("|---------|-----------|-----------|---------|");

    for (size, label) in payload_sizes.iter().zip(size_labels.iter()) {
        let r = benchmark_copy_vs_zero_copy(*size, NUM_ITERATIONS);
        println!(
            "| {:7} | {:7.2} ms | {:7.2} ms | {:5.1}x  |",
            label, r.copy_time_ms, r.zero_copy_time_ms, r.speedup
        );
    }

    // When to use zero-copy
    println!("\n--- When to Use Zero-Copy ---\n");
    println!("Recommended when:");
    println!("  - Payload size > 64 KB");
    println!("  - Same-host communication (intra-process or inter-process)");
    println!("  - High message rates with large payloads");
    println!("  - CPU is bottleneck (reduces memcpy overhead)\n");

    println!("Not recommended when:");
    println!("  - Small payloads (< 1 KB) - overhead dominates");
    println!("  - Cross-network communication (copy required anyway)");
    println!("  - Security isolation required between processes");

    // Memory considerations
    println!("\n--- Memory Considerations ---\n");
    println!("Shared memory must be configured:");
    println!("  - /dev/shm size (Linux): check with 'df -h /dev/shm'");
    println!("  - Segment size: must fit all loaned samples");
    println!("  - Cleanup: segments persist until explicitly removed");

    // Rust-specific notes
    println!("\n--- Rust-Specific Notes ---\n");
    println!("Rust's ownership model works well with zero-copy:");
    println!("  - loan_sample() transfers ownership to user");
    println!("  - write_loaned() transfers ownership back to DDS");
    println!("  - Borrow checker ensures safe concurrent access\n");
    println!("For shared memory in Rust:");
    println!("  - shared_memory crate for cross-process");
    println!("  - memmap2 crate for memory-mapped files");

    println!("\n=== Sample Complete ===");
    Ok(())
}
